swift_browser_ui_frontend/src/common/store.js

```diff
@@ -145,78 +142,90 @@ const store = new Vuex.Store({
     },
   },
   actions: {
-    updateContainers: async function ({ commit, dispatch }, signal) {
+    updateContainers: async function (
+      { commit, dispatch, state },
+      {route, signal},
+    ) {
       commit("loading", true);
-      let containers = [];
-      await getBuckets().then((ret) => {
-        if (ret.status != 200) {
+      let project = route.params.project;
+      commit("updateContainers", []);
+      let containers;
+      let marker = "";
+      do {
+        containers = [];
+        containers = await getBuckets(
+          project,
+          marker,
+        ).catch(() => {
           commit("loading", false);
-        }
-        containers = ret;
-        commit("updateContainers", ret);
-        commit("loading", false);
-      }).catch(() => {
+        });
         commit("loading", false);
+        if (containers.length > 0) {
+          commit("updateContainers", state.containerCache.concat(containers));
+          marker = containers.pop().name;
+        }
+      } while (containers.length > 0);
```

Comment:
I didn't get this. It looks like it is making an extra request for each container. Why does it do that?

Reply:
"Null-terminated" listings. If there's nothing left after the marker, Swift API returns an empty listing signifying the end. It can't reliably be done by checking if the page contains less items than the page size, the page size is an upper limit and is not always an exact thing. Openstack's own implementation does it in a similar manner.